### PR TITLE
Fix codegen for rvalue aggregate raw pointer to an adt with slice tail

### LINKED ISSUE #3638: ICE: Kani compiler crashes due to incorrect handling of Adt with slice tail
I tried this code:

```rust
// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT

#![feature(ptr_metadata)]

use std::ptr::NonNull;

#[derive(kani::Arbitrary)]
struct Wrapper<T: ?Sized>(usize, T);

#[cfg(kani)]
#[kani::proof]
fn main() {
    // Create a SampleTrait object from SampleStruct
    let original: Wrapper<[u8; 10]> = kani::any();
    let slice: &Wrapper<[u8]> = &original;

    // Get the raw data pointer and metadata for the trait object
    let slice_ptr = NonNull::new(slice as *const _ as *mut ()).unwrap();
    let metadata = std::ptr::metadata(slice);

    // Create NonNull<dyn SampleTrait> from the data pointer and metadata
    let nonnull: NonNull<Wrapper<[u8]>> =
        NonNull::from_raw_parts(slice_ptr, metadata);

}

```

using the following command line invocation:

```
kani slice_tail.rs
```

with Kani version: 0.56.0

I expected to see this happen: verification succeed

Instead, this happened: Kani compiler crashed

<pre>$ kani adt_slice.rs 
Kani Rust Verifier 0.56.0 (standalone)
thread &apos;rustc&apos; panicked at /tmp/workspace/kani-project/kani/cprover_bindings/src/goto_program/expr.rs:733:9:
Can&apos;t apply .member operation to
	Expr { value: Symbol { identifier: &quot;_RINvNtNtCsfIIKBCXq9GN_4core3ptr8metadata18from_raw_parts_mutINtCsgncW3vUYF2n_9adt_slice7WrapperShEuEBZ_::1::var_2::metadata&quot; }, typ: CInteger(SizeT), location: None, size_of_annotation: None }
	_vtable_ptr
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Kani unexpectedly panicked during compilation.
Please file an issue here: https://github.com/model-checking/kani/issues/new?labels=bug&amp;template=bug_report.md

[Kani] current codegen item: codegen_function: std::ptr::from_raw_parts_mut::&lt;Wrapper&lt;[u8]&gt;, ()&gt;
_RINvNtNtCsfIIKBCXq9GN_4core3ptr8metadata18from_raw_parts_mutINtCsgncW3vUYF2n_9adt_slice7WrapperShEuEBZ_
[Kani] current codegen location: Loc { file: &quot;/tmp/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/metadata.rs&quot;, function: None, start_line: 125, start_col: Some(1), end_line: 128, end_col: Some(12), pragmas: [] }
<font color="#A2734C"><b>warning</b></font><b>: 2 warnings emitted</b>

<font color="#C01C28"><b>error:</b></font> <b>/tmp/workspace/kani-project/kani/target/kani/bin/kani-compiler exited with status exit status: 101</b></pre>

### PR DESCRIPTION
**Pull request by carolynzech**
Fix the codegen for a raw pointer to a dynamically-sized ADT. Traverse the type until we find the tail, which will either be a trait object or a slice, and generate the pointer accordingly.

Resolves #3638, #3615

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — CHANGES_REQUESTED** [review]
You might need to traverse the type, since the tail could be multiple levels down. E.g.:

```rust
Wrapper<Wrapper<[u8]>>
```

**Comment by celinval** [comment]
I pushed a new patch to this PR. I pulled the code from https://github.com/model-checking/kani/pull/3687. Note that I already added the entire `abi` module that is used in the original PR. To avoid warnings, I added an annotation `#[allow(dead_code)]` to the new structure for now.

File: kani-compiler/src/kani_middle/abi.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    /// For foreign types, return None.
+    /// For unsized types, this should return either a slice, a string slice, a dynamic type.
+    /// For other types, this function will return `None`.
+    pub fn unsized_tail(&self) -> Option<Ty> {
+        if self.layout.is_unsized() {
+            match self.ty.kind().rigid().unwrap() {
+                RigidTy::Slice(..) | RigidTy::Dynamic(..) | RigidTy::Str => Some(self.ty),
+                RigidTy::Adt(..) | RigidTy::Tuple(..) => {
+                    // Recurse the tail field type until we find the unsized tail.
+                    self.last_field_layout().unsized_tail()
```
Comment (qinheping) [inline kani-compiler/src/kani_middle/abi.rs:61]: Are unsized fields always the last fileds?

File: tests/kani/AggregateRvalue/slice_ptr.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Test codegen for a raw pointer to a struct whose last field is a slice
+
+#![feature(layout_for_ptr)]
+#![feature(ptr_metadata)]
+
+// https://github.com/model-checking/kani/issues/3615
+mod issue_3615 {
```
Comment (qinheping) [inline tests/kani/AggregateRvalue/slice_ptr.rs:9]: Can this PR also fix #2256?

File: kani-compiler/src/kani_middle/abi.rs (hunk 1 above)
Comment (celinval) [reply kani-compiler/src/kani_middle/abi.rs:61]: Yes. See https://doc.rust-lang.org/std/mem/fn.size_of_val_raw.html

File: kani-compiler/src/kani_middle/abi.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                }
+                RigidTy::Foreign(_) => Some(self.ty),
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/kani_middle/abi.rs:63]: No, I should update the comment instead. :grimacing:

File: kani-compiler/src/kani_middle/abi.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        self.unsized_tail().map_or(false, |tail| {
+            let kind = tail.kind();
+            kind.is_slice() | kind.is_str()
+        })
+    }
+
+    /// Return the unsized tail of the type if this is an unsized type.
+    ///
+    /// For foreign types, return None.
+    /// For unsized types, this should return either a slice, a string slice, a dynamic type.
```
Comment (celinval) [inline kani-compiler/src/kani_middle/abi.rs:53]:
```suggestion
    /// For unsized types, this should return either a slice, a string slice, a dynamic type,
    /// or a foreign type.
```

File: kani-compiler/src/kani_middle/abi.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                }
+                RigidTy::Foreign(_) => Some(self.ty),
+                _ => unreachable!("Expected unsized type but found `{}`", self.ty),
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Return the type of the elements of the array or slice at the unsized tail of this type.
```
[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/kani_middle/abi.rs:71]:
```suggestion
    /// Return the type of the elements of the slice or `str` at the unsized tail of this type.
```

File: tests/kani/AggregateRvalue/slice_ptr.rs (hunk 2 above)
Comment (celinval) [reply tests/kani/AggregateRvalue/slice_ptr.rs:9]: Oh, that's likely going to fix it. I'll check it later, and add the test here if it does. Thanks!

**Review by qinheping — APPROVED** [review]
LGTM

**Pull request merged**